### PR TITLE
Switch Persistence [4/4]: Switch Integration and Test Suite

### LINKED ISSUE #80: Persist active onion circuits before forwarding
## Issue 

Currently within the daemon, the `htlcSwitch` manages the teardown/setup of [onion-circuits](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch.go#L66) generated during the forwarding of multi-hop payments. However, as implemented currently, a well timed daemon restart may lead to possibly loss of funds. This is due the fact that the onion circuits are currently kept in [volatile memory](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch.go#L135), and not persisted to disk before forwarding. 

Onion circuits are uniquely defined with the following struct: 
```go
type onionCircuit struct {
    rHash         [32]byte
    clearNode  *btcec.PublicKey
    settleNode *btcec.PublicKey
}
```

where: 
  * `clearNode` is the node that originally forwarded the HTLC to us, clearing it on the incoming channel
  * and `settleNode` is the node we sent the outgoing HTLC to

Once the `settleNode` sends us an HTLC settle message, we the switch needs to forward the settle message to the `clearNode` in order to claim the funds and in the process gain the fee related to the HTLC. 

## Steps to Completion 

Persist the `htlcManager's` `paymentCircuits` map to disk. Ideally, the map should be removed all together and modifications to the prior map should instead take the form of disk accesses in order to side-step any potential consistency issues related to maintain the state both in memory and on disk.

### LINKED ISSUE #304: Implement persistent HTLC expiry decayed Sphinx packet replay protection
# Issues 

[Our implementation](https://github.com/lightningnetwork/lightning-onion) of [Lighting's onion routing protocol](https://github.com/lightningnetwork/lightning-rfc/blob/master/04-onion-routing.md) critically depends on proper replay protection in order to deliver on its privacy guarantees. Otherwise, an attacker can inject a prior HTLC into the network to attempt to re-trace its path in the case it has compromised one or more nodes within a potential path. 

Currently, we implement a basic version of replay protection backed by [non-persistent storage](https://github.com/lightningnetwork/lightning-onion/blob/master/sphinx.go#L677) within the `sphinx.Router` struct. This insufficient as if the daemon restart, then it will be susceptible to propagated replays as it's forgotten all the prior shared secrets. Additionally, as is now this storage is never reclaimed. We can reclaim this storage as HTLC's have a set _expiration_ date, past that expiration date we no longer need to remember the shared secret once an HTLC as expired as [we'll reject it within the state machine if it's already expired](https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L1186). 

We'll need to implement a persistent, decaying shared-secret log in order to properly implement replay protection. 

# Steps to Completion 

- [ ] Reject HTLC's which expire a period deemed too far in the future. We already ensure that they exactly meet our time lock delta parameters, but this measure will ensure that we don't even accept such HTLC's. This is critical as otherwise this presents a DoS vector against the persistent shared secret log. 

- [ ] Create a new on-disk data-structure implementing a persistent, decayed shared secret log. This should likely stand behind a new interface created that will replace the current [`sharedSecret map`](https://github.com/lightningnetwork/lightning-onion/blob/master/sphinx.go#L610) in the `sphinx.Router`. 

    We have two options here:

    1. Re-use `boltdb` for the persistent log. A struct wrapping `channeldb` would implement to to-be-created interface in the `lightning-onion` repo. This is the easiest path forward likely but would increase write contention as we'd need an additional write for each HTLC processed, though they could be batched. 
    2. Implement an on-disk replay protection log for _each_ channel implemented as a flat file. This would involved linearly scanning the file, but periodic checkpoints which contain skip pointers to later sections would allow searches to be short circuited. 

    In either case, the format of the on-disk structure would be: `(sharedHash, expiryHeight)`. Within the `sphinx.Router` we'd consult this rather than the current on-disk map. Where `sharedSecretHash = sha256(sharedSecret)[:20])`. Rather than writing the raw shared secret to disk, we instead write a truncated sha256 hash of it. By doing this, we avoid writing sensitive data (the shared secret) to disk, and also save a bit of space by truncating to 20 bytes instead of 32 bytes. 

- [ ] Create a persistent goroutine that garbage collects entries within the log. When `currentHeight > expiryHeight`, the entry can safely be removed from the on-disk storage.

### PR DESCRIPTION
**Pull request by cfromknecht**
This the final component of the switch persistence saga, which building on #775, #777, and #791, hooking up all of the subcomponents including sphinx replay protection, forwarding packages, and the persistence circuit map. The primary things included in this PR:

- Internal replay protection/handling of forwarded HTLCs
- Replayable mailboxes that supercede the lifecycle of a link and maintained by the switch.
- Persistent payment sequencer, to ensure that locally-initiated payment circuits do not collide across restarts.
- Batched replay protection for sphinx packets and shared secret decryption in #775
- Batched modification of the circuit map changes in #791.
- Batched writing/modifying of fwdpkgs, done so atomically when receiving revocations or signing commitment states.
- Garbage collection of fwdpkgs using a ticker.

NOTE: With this PR, we WILL accept duplicate payments for the same invoice. In reality, it is the sender's job to ensure that this never happens as someone could modify the code to accept a duplicate payment anyway.

NOTE: This PR doesn't compile on it's own, see #761 for full integration with other PRs and passing test cases.

Assuming #775 is merged prior, this closes #304.
Assuming #791 is merged prior, this closes #80.

### CODE REVIEW

File: htlcswitch/switch.go [hunk 1]
```diff
@@ -358,7 +412,184 @@ func (s *Switch) updateLinkPolicies(c *updatePoliciesCmd) error {
...
+
+		// We don't handle the error here since this method always
+		// returns an error.
+		s.failAddPacket(source, packet, failure, addErr)
+	}
+
+	return errChan
+}
+
+func (s *Switch) proxyFwdErrs(num *int, wg *sync.WaitGroup,
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Missing `godoc` comment.

File: htlcswitch/switch.go [hunk 2]
```diff
@@ -463,83 +710,92 @@ func (s *Switch) handleLocalDispatch(packet *htlcPacket) error {
...
-				userErr := fmt.Sprintf("unable to de-obfuscate onion failure, "+
-					"htlc with hash(%x): %v", payment.paymentHash[:], err)
-				log.Error(userErr)
-				failure = &ForwardingError{
-					ErrorSource:    s.cfg.SelfKey,
-					ExtraMsg:       userErr,
-					FailureMessage: lnwire.NewTemporaryChannelFailure(nil),
-				}
-			}
+func (s *Switch) parseFailedPayment(payment *pendingPayment, pkt *htlcPacket,
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Missing `godoc` comment.

File: htlcswitch/sequencer.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		return nil, err
+	}
+
+	return g, nil
+}
+
+// NextID returns a unique sequence number for every invocation, persisting the
+// assignment to avoid reuse.
+func (g *persistentSequencer) NextID() (uint64, error) {
+	var nextPaymentID uint64
```
Comment (Roasbeef) [inline htlcswitch/sequencer.go]: Not a blocker, but I think we should add a TODO for you suggestion of holding a set of payment ID's (like 10 or so), and allocating them in batches. This way we can batch the reads/writes for all pending payments.

File: htlcswitch/link.go [hunk 4]
```diff
@@ -525,77 +575,210 @@ func (l *channelLink) syncChanStates() error {
...
+			for _, fwdPkg := range fwdPkgs {
+				if fwdPkg.State != channeldb.FwdStateCompleted {
+					continue
+				}
+
+				err = l.channel.RemoveFwdPkg(fwdPkg.Height)
+				if err != nil {
+					l.warnf("unable to remove fwd pkg "+
+						"for height=%d: %v",
+						err, fwdPkg.Height)
```
Comment (cfromknecht) [inline htlcswitch/link.go]: Note to self, switch the arguments

File: htlcswitch/link.go [hunk 5]
```diff
@@ -1530,23 +1954,34 @@ func (l *channelLink) processLockedInHtlcs(
...
-					log.Warnf("Rejecting duplicate "+
+				// If the invoice is already settled, we choose
+				// to accept the payment to simplify failure
+				// recovery.
+				//
+				// NOTE: Though our recovery and forwarding logic is
+				// predominately batched, settling invoices
+				// happens iteratively. We may reject one of of
+				// two payments for the same rhash at first, but
+				// then restart and settle both after seeing
```
Comment (cfromknecht) [inline htlcswitch/link.go]: reject both after*

File: htlcswitch/link.go [hunk 6]
```diff
@@ -120,18 +124,33 @@ type ChannelLinkConfig struct {
...
-	//
-	// TODO(roasbeef): remove in favor of simple ForwardPacket closure func
+	// Circuits provides restricted access to the switch's circuit map,
+	// allowing the link to open and close circuits.
+	Circuits CircuitModifier
+
+	// Switch provides a reference to the HTLC switch, we only use this in
+	// testing to access circuit operations not typically exposed by the
+	// CircuitModifier.
+	// TODO(conner): remove after refactoring htlcswitch testing framework.
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: `CircuitModifier` is a subset of the interface exposed by the `CircuitMap`, but limits the scope to those methods used by each link

File: htlcswitch/link.go [hunk 7]
```diff
@@ -120,18 +124,33 @@ type ChannelLinkConfig struct {
...
 	Switch *Switch
 
+	// ForwardPackets attempts to forward the batch of htlcs through the
+	// switch. Any failed packets will be returned to the provided
+	// ChannelLink.
+	ForwardPackets func(ChannelLink, ...*htlcPacket) chan error
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed this by having fails be added directly back into the mailboxes, so no need to pass the link anymore

File: htlcswitch/link.go [hunk 8]
```diff
@@ -525,77 +582,210 @@ func (l *channelLink) syncChanStates() error {
...
-		if !ok {
-			// If this isn't a settle message, then we'll skip it.
-			continue
-		}
-
-		// Otherwise, we'll note the ID of the HTLC we're settling so we
-		// don't duplicate it below.
-		htlcsSettled[settleMsg.ID] = struct{}{}
-	}
+	// TODO(conner): uncomment/rework this section after further integration
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Cool, I'll go ahead and remove this

File: htlcswitch/link.go [hunk 9]
```diff
@@ -525,77 +582,210 @@ func (l *channelLink) syncChanStates() error {
...
+	for {
+		select {
+		case <-fwdPkgGcTick:
+			fwdPkgs, err := l.channel.LoadFwdPkgs()
+			if err != nil {
+				l.warnf("unable to load fwdpkgs for gc: %v", err)
+				continue
+			}
+
+			for _, fwdPkg := range fwdPkgs {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed.

File: htlcswitch/link.go [hunk 10]
```diff
@@ -627,11 +835,39 @@ func (l *channelLink) htlcManager() {
...
+		return
+	}
+
+	// With our link's in-memory state fully reconstructed, spawn a
+	// goroutine to manage the reclamation of disk space occupied by
+	// completed forwarding packages.
+	l.wg.Add(1)
+	go l.fwdPkgGarbager()
+
+	// TODO(roasbeef): check to see if able to settle any currently pending
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed

File: htlcswitch/link.go [hunk 11]
```diff
@@ -1374,22 +1735,25 @@ func (l *channelLink) updateChannelFee(feePerKw lnwallet.SatPerKWeight) error {
...
 }
 
-// processLockedInHtlcs serially processes each of the log updates which have
-// been "locked-in". An HTLC is considered locked-in once it has been fully
-// committed to in both the remote and local commitment state. Once a channel
-// updates is locked-in, then it can be acted upon, meaning: settling HTLCs,
-// cancelling them, or forwarding new HTLCs to the next hop.
-func (l *channelLink) processLockedInHtlcs(
-	paymentDescriptors []*lnwallet.PaymentDescriptor) []*htlcPacket {
+func (l *channelLink) processRemoteSettleFails(fwdPkg *channeldb.FwdPkg,
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed

File: htlcswitch/link.go [hunk 12]
```diff
@@ -1374,22 +1735,25 @@ func (l *channelLink) updateChannelFee(feePerKw lnwallet.SatPerKWeight) error {
...
+	if len(settleFails) == 0 {
+		return
+	}
+
+	log.Debugf("ChannelLink(%v): settle-fail-filter %v",
+		l.ShortChanID(), fwdPkg.SettleFailFilter)
+
+	var switchPackets []*htlcPacket
+	for i, pd := range settleFails {
+
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed

File: htlcswitch/switch.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: Fixed

File: htlcswitch/switch.go [hunk 13]
```diff
@@ -358,7 +412,184 @@ func (s *Switch) updateLinkPolicies(c *updatePoliciesCmd) error {
...
+	// responses until this function returns to the user.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Done()
+
+	// Spawn a goroutine the proxy the errs back to the returned err chan.
+	// This is done to ensure the err chan returned to the caller closed
+	// properly, alerting the receiver of completion or shutdown.
+	s.wg.Add(1)
+	go s.proxyFwdErrs(&numSent, &wg, fwdChan, errChan)
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: I believe `ErrGroup` will return as soon as one error occurs, whereas the use case here is simply to ensure that all errors are propagated and logged in the channel. If we select on the `errChan` in addition to a quit, it's possible that some error messages can be lost during shutdown. This ensures that all of them will be logged before exiting.

File: htlcswitch/switch.go [hunk 14]
```diff
@@ -358,7 +412,184 @@ func (s *Switch) updateLinkPolicies(c *updatePoliciesCmd) error {
...
+			failedPackets = append(failedPackets, packet)
+			fails = fails[1:]
+		}
+	}
+
+	// Now, forward any packets for circuits that were successfully added to
+	// the switch's circuit map.
+	for _, packet := range addedPackets {
+		s.routeAsync(packet, fwdChan)
+		numSent++
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: The wait group should only unblock after the outer function exits, which occurs after all possible packets have been forwarded or failed. I think this should be okay?

File: htlcswitch/switch.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: Fixed.

File: htlcswitch/switch.go [hunk 15]
```diff
@@ -979,8 +1307,15 @@ func (s *Switch) Stop() error {
...
 
 	close(s.quit)
+
+	for _, mailBox := range s.mailboxes {
+		mailBox.Stop()
+	}
+
 	s.wg.Wait()
 
+	s.cfg.DB.Close()
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: Fixed, I think you're correct. May need to update the tests to properly close the db

File: htlcswitch/switch.go [hunk 16]
```diff
@@ -1043,6 +1386,19 @@ func (s *Switch) addLink(link ChannelLink) error {
 	return nil
 }
 
+func (s *Switch) getOrCreateMailBox(chanID lnwire.ShortChannelID) MailBox {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: Fixed.

File: htlcswitch/switch.go (hunk 1 above)
Comment (cfromknecht) [reply htlcswitch/switch.go]: Fixed.

File: htlcswitch/switch.go (hunk 2 above)
Comment (cfromknecht) [reply htlcswitch/switch.go]: Fixed.

File: htlcswitch/sequencer.go (hunk 3 above)
Comment (cfromknecht) [reply htlcswitch/sequencer.go]: Fixed!

File: htlcswitch/link.go (hunk 4 above)
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed.

File: htlcswitch/link.go (hunk 5 above)
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed

File: htlcswitch/switch.go [hunk 17]
```diff
@@ -565,46 +852,22 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
 	// payment circuit within our internal state so we can properly forward
 	// the ultimate settle message back latter.
 	case *lnwire.UpdateAddHTLC:
-		if packet.incomingChanID == (lnwire.ShortChannelID{}) {
-			// A blank incomingChanID indicates that this is a
-			// pending user-initiated payment.
+		if packet.incomingChanID == sourceHop {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply htlcswitch/switch.go]: Defined in another file, basically means that it was a user initiated payment.

[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: It's an empty `lnwire.ShortChannelID`, just like `exitHop`!

File: htlcswitch/link.go [hunk 18]
```diff
@@ -525,77 +575,124 @@ func (l *channelLink) syncChanStates() error {
...
+		needUpdate = l.processRemoteAdds(fwdPkg, adds)
+	}
+
+	return needUpdate, nil
+}
+
+// fwdPkgGarbager periodically reads all forwarding packages from disk and
+// removes those that can be discarded. It is safe to do this entirely in the
+// background, since all state is coordinated on disk. This also ensures the
+// link can continue to process messages and interleave database accesses.
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: Fixed

File: htlcswitch/switch.go [hunk 19]
```diff
@@ -372,6 +444,190 @@ func (s *Switch) updateLinkPolicies(c *updatePoliciesCmd) error {
...
+	fwdChan, errChan chan error) {
+	defer s.wg.Done()
+	defer func() {
+		close(errChan)
+	}()
+
+	// Wait here until the outer function has finished persisting
+	// and routing the packets. This guarantees we don't read from num until
+	// the value is accurate.
+	wg.Wait()
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: Nice call, that is definitely cleaner

[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: I think this would also allow us to pass `num` by value instead of by reference

File: htlcswitch/switch.go [hunk 20]
```diff
@@ -832,12 +1217,16 @@ func (s *Switch) htlcForwarder() {
 			}
 		}
 
+		log.Infof("flushing fwding events")
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Looks like a stray debugging statement.

File: htlcswitch/switch.go [hunk 21]
```diff
@@ -832,12 +1217,16 @@ func (s *Switch) htlcForwarder() {
...
+
 		// Before we exit fully, we'll attempt to flush out any
 		// forwarding events that may still be lingering since the last
 		// batch flush.
 		if err := s.FlushForwardingEvents(); err != nil {
 			log.Errorf("unable to flush forwarding events: %v", err)
 		}
+
+		log.Infof("fwding events flushed")
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Looks like a stray debugging statement.

File: htlcswitch/switch.go [hunk 22]
```diff
@@ -1029,9 +1421,139 @@ func (s *Switch) Start() error {
...
 
+	if err := s.reforwardResponses(); err != nil {
+		log.Errorf("unable to reforward responses: %v", err)
+		return err
+	}
+
 	return nil
 }
 
+func (s *Switch) reforwardResponses() error {
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Missing `godoc` comment.

File: htlcswitch/switch.go [hunk 23]
```diff
@@ -1029,9 +1421,139 @@ func (s *Switch) Start() error {
...
+			return err
+		}
+
+		s.reforwardSettleFails(fwdPkgs)
+	}
+
+	return nil
+}
+
+func (s *Switch) loadChannelFwdPkgs(
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Missing `godoc` comment.

File: htlcswitch/switch.go [hunk 24]
```diff
@@ -1029,9 +1421,139 @@ func (s *Switch) Start() error {
...
+		)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+
+	return fwdPkgs, nil
+}
+
+func (s *Switch) reforwardSettleFails(fwdPkgs []*channeldb.FwdPkg) {
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: Missing `godoc` comment.

File: htlcswitch/switch_test.go [hunk 25]
```diff
@@ -11,39 +13,706 @@ import (
...
-	hash2, _ = chainhash.NewHash(bytes.Repeat([]byte("b"), 32))
+func genPreimage() ([32]byte, error) {
+	var preimage [32]byte
+	if _, err := io.ReadFull(rand.Reader, preimage[:]); err != nil {
+		return preimage, err
+	}
+	return preimage, nil
+}
+
+// TestSwitchForward checks the ability of htlc switch to forward add/settle
```
Comment (Roasbeef) [inline htlcswitch/switch_test.go]: Really dig these expanded set of test cases for the switch!

File: lnd_test.go [hunk 26]
```diff
@@ -1973,6 +1973,179 @@ func testChannelForceClosure(net *lntest.NetworkHarness, t *harnessTest) {
...
+	// he does fail, it is almost certainly caused by the sphinx replay
+	// protection, as it is the only validation we do in hodl mode.
+	resp, err := payStream.Recv()
+	if err != nil {
+		t.Fatalf("unable to receive payment response: %v", err)
+	}
+
+	// Construct the response we expect after sending a duplicate packet
+	// that fails due to sphinx replay detection.
+	replayErr := fmt.Sprintf("unable to route payment to destination: "+
```
Comment (Roasbeef) [inline lnd_test.go]: Not a blocker, but in the future we may want to make the string comparison here a bit looser, or just switch to precise error codes in the long term.

File: htlcswitch/switch.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/switch.go]: Ended up reverting because there was some reliability issues, can circle back on this.

File: lnd_test.go (hunk 26 above)
Comment (cfromknecht) [reply lnd_test.go]: Yep definitely :)

**Pull request closed without merging**